README.md

---
# Get Started
```
$ git clone https://github.com/tddey01/switfs-auth.git
$ export GOPROXY=https://goproxy.io,direct
$ export GO111MODULE=on
$ make

$ sophon-auth
```

# RESTFul API
## 1. verify token
- method: POST
- route : http://localhost:8989/verify

- Body params:

name | type | desc |e.g.
---|---|---|---
token | string| jwt token | eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiUmVubmJvbiIsInBlcm0iOiJhZG1pbiIsImV4dCI6ImV5SkJiR3h2ZHlJNld5SnlaV0ZrSWl3aWQzSnBkR1VpTENKemFXZHVJaXdpWVdSdGFXNGlYWDAifQ.gONkC1v8AuY-ZP2WhU62EonWmyPeOW1pFhnRM-Fl7ko

- response
```
# status 200 :
{
    "name": "Rennbon",
    "perm": "admin",
    "ext": "eyJBbGxvdyI6WyJyZWFkIiwid3JpdGUiLCJzaWduIiwiYWRtaW4iXX0"
}
# status 401:
{
    "error": "A non-registered token"
}
```

## 2. generate token
- method: POST
- route : http://localhost:8989/genToken
- Body params:

name | type | desc |e.g.
---|---|---|---
name | string| The name of the description |  Rennbon
perm | string | admin,sign,write,read | admin
extra | string | custom payload | 
- response
```
# status 200 :
{
    "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiUmVubmJvbiIsInBlcm0iOiJhZG1pbiIsImV4dCI6ImV5SkJiR3h2ZHlJNld5SnlaV0ZrSWl3aWQzSnBkR1VpTENKemFXZHVJaXdpWVdSdGFXNGlYWDAifQ.gONkC1v8AuY-ZP2WhU62EonWmyPeOW1pFhnRM-Fl7ko"
}

```

## 3. remove token
- method: DELETE
- route : http://localhost:8989/token
- Body params:

name | type | desc |e.g.
---|---|---|---
token | string| jwt token |  eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiUmVubmJvbiIsInBlcm0iOiJhZG1pbiIsImV4dCI6ImV5SkJiR3h2ZHlJNld5SnlaV0ZrSWl3aWQzSnBkR1VpTENKemFXZHVJaXdpWVdSdGFXNGlYWDAifQ.gONkC1v8AuY-ZP2WhU62EonWmyPeOW1pFhnRM-Fl7ko

- response
```
# status 200 
```

## 4. list token info
- method: GET
- route : http://localhost:8989/tokens

name | type | desc |e.g.
---|---|---|---
skip | int | \>= 0  |  1
limit | int | \> 0 | 20
- response
```
# status 200 
[
    {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiUmVubmJvbiIsInBlcm0iOiJhZG1pbiIsImV4dCI6ImV5SkJiR3h2ZHlJNld5SnlaV0ZrSWl3aWQzSnBkR1VpTENKemFXZHVJaXdpWVdSdGFXNGlYWDAifQ.Ct8Lc-lc1nppIejRz-y0ht7yAnzB0-bpwk4Vkk0k-TM",
        "name": "Rennbon",
        "createTime": "2021-03-30T17:02:32.347018+08:00"
    },
    {
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoibG90dXMtbWluZXIiLCJwZXJtIjoiYWRtaW4iLCJleHQiOiJleUpCYkd4dmR5STZXeUp5WldGa0lpd2lkM0pwZEdVaUxDSnphV2R1SWl3aVlXUnRhVzRpWFgwIn0.cwK2GgDydEY8pC8NBW2wlOBaoxDZhIdA1xgV6WSF63g",
        "name": "lotus-miner",
        "createTime": "2021-04-01T15:57:39.858826+08:00"
    }
]
```
---

# CLI
## 1. generate token
```
# show help
$ ./sophon-auth token gen -h
USAGE:
   sophon-auth token gen [command options] [name]

OPTIONS:
   --perm value   permission for API auth (read, write, sign, admin) (default: "read")
   --extra value  custom string in JWT payload

$ ./sophon-auth token gen token1 --perm admin --extra custom_str
generate token success: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoidG9rZW4xIiwicGVybSI6InJlYWQiLCJleHQiOiIifQ.s3jvO-yewsf3PHMF-tsWSbb-3aW7V-tlMsnEAkYdxgA
```
## 2. list token info
```
# show help

$ ./sophon-auth token list -h
USAGE:
   sophon-auth token list [command options] [arguments...]

OPTIONS:
   --skip value   (default: 0)
   --limit value  (default: 20)
   --help, -h     show help (default: false)

$ ./sophon-auth token list --skip 0 --limit 10
num     name          perm    createTime              token
1       token1        admin   2021-05-31 18:45:02     eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoidG9rZW4xIiwicGVybSI6InJlYWQiLCJleHQiOiIifQ.s3jvO-yewsf3PHMF-tsWSbb-3aW7V-tlMsnEAkYdxgA
2       token2        read    2021-06-18 13:31:47     eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiZmF0bWFuMTMiLCJwZXJtIjoicmVhZCIsImV4dCI6IiJ9.F0frWmZSsEpyZIY_VOQ9WiAVxAfzqUdhvrU16ltbP9U
3       token3        write   2021-06-19 00:14:02     eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiZmF0bWFuMTMiLCJwZXJtIjoid3JpdGUiLCJleHQiOiIifQ.Txu3yYCAtbKL9jSzsf3ldDWz7WX5F3w7RnQBDzMtY-0
4       token4        sign    2021-07-06 11:14:06     eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoiaGUiLCJwZXJtIjoicmVhZCIsImV4dCI6IiJ9.Hjmnh4snGYc1lT2PplH4tffIdBNta7QPRiwCeWsty2s
```

## 3. remove token
```
# show help
$ ./sophon-auth token rm -h
USAGE:
   sophon-auth token rm [command options] [token]

OPTIONS:
   --help, -h  show help (default: false)

$ ./sophon-auth token rm eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoidG9rZW4xIiwicGVybSI6InJlYWQiLCJleHQiOiIifQ.s3jvO-yewsf3PHMF-tsWSbb-3aW7V-tlMsnEAkYdxgA 
remove token success: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJuYW1lIjoidG9rZW4xIiwicGVybSI6InJlYWQiLCJleHQiOiIifQ.s3jvO-yewsf3PHMF-tsWSbb-3aW7V-tlMsnEAkYdxgA

```
# Config
>the default config path is "~/.auth-auth/config.toml"
```
Listen = "127.0.0.1:8989"
ReadTimeout = "1m"
WriteTimeout = "1m"
IdleTimeout = "1m"

[db]
  # support: badger (default), mysql
  # the mysql DDL is in the script package
  type = "badger"
  # The following parameters apply to MySQL
  DSN = "rennbon:111111@(127.0.0.1:3306)/auth_server?parseTime=true&loc=Local&charset=utf8mb4&collation=utf8mb4_unicode_ci&readTimeout=10s&writeTimeout=10s"
  # conns 1500 concurrent
  maxOpenConns = 64
  maxIdleConns = 128
  maxLifeTime = "120s"
  maxIdleTime = "30s"

[log]
  # trace,debug,info,warning,error,fatal,panic
  # output level
  logLevel = trace
  # db type, 1:influxDB
  type = 1
  # db hook switch
  hookSwitch = true

[Trace]
  # Enable trace
  JaegerTracingEnabled = true
  # Frequency of collection
  ProbabilitySampler = 1.0
  JaegerEndpoint = "127.0.0.1:6831"
  ServerName = "sophon-auth"
```

## [Script](./script)
- influxdb-docker-compose.yml => rename docker-compose.yml and install influxdb in docker
- influxDB_view.md => histogram and graph view config


CHANGELOG.md
# sophon-auth changelog

## v1.11.0

* bump up version to v1.11.0

## v1.11.0-rc1

### New Features
* feat: add status api to detect api ready / 添加状态检测接口 [[#144](https://github.com/tddey01/switfs-auth/pull/144)]
* feat: use thirty party healthcheck lib  / 添加healthcheck接口 [[#145](https://github.com/tddey01/switfs-auth/pull/145)]
* feat: add api protection / 增加接口保护  [[#140](https://github.com/tddey01/switfs-auth/pull/140)]
* feat: return detailed error infor after authentication failure  /鉴权失败后返回更加详细的错误信息 [[#160](https://github.com/tddey01/switfs-auth/pull/160)]
* feat: add docker push / 增加推送到镜像仓库的功能 [[#161](https://github.com/tddey01/switfs-auth/pull/161)]

### Bug Fixes

* fix: repo not exist by 修复启动时目录不存在从而启动失败的问题 [[#157](https://github.com/tddey01/switfs-auth/pull/157)]
* fix: not set flag value to config  /修复配置错误 [[#158](https://github.com/tddey01/switfs-auth/pull/158)]
* fix: cli not found config  / 修复创建目录失败的问题 [[#159](https://github.com/tddey01/switfs-auth/pull/159)]


## v1.10.1

* 查询参数为空时不重置请求url [[#153](https://github.com/tddey01/switfs-auth/pull/153)]
* 补充对 delegated 地址的支持 [[#154](https://github.com/tddey01/switfs-auth/pull/154)]

## v1.10.0

* 升级 go-jsonrpc 到 v0.1.7

## v1.10.0-rc2

* 简化 authClient 接口，并增加 context [[#126](https://github.com/tddey01/switfs-auth/pull/126)]
* 重写 url 中的地址参数 [[#127](https://github.com/tddey01/switfs-auth/pull/127)]
* 增加用户数据隔离的工具 [[#130](https://github.com/tddey01/switfs-auth/pull/130)]
* 调整 jwtclient.IAtuhClient 接口 [[#137](https://github.com/tddey01/switfs-auth/pull/137)]

## v1.10.0-rc1

* github action 增加 dispatch 事件 [[#138](https://github.com/tddey01/switfs-auth/pull/138)]
* 升级 github.com/gin-gonic/gin 版本，从 1.7.0 升级到 1.7.7 [[#139](https://github.com/tddey01/switfs-auth/pull/139)]
* 升级 github.com/prometheus/client_golang 版本，从 1.11.0 升级到 1.11.1 [[#141](https://github.com/tddey01/switfs-auth/pull/141)]


cli/token.go
package cli

import (
	"fmt"

	"github.com/urfave/cli/v2"

	"github.com/tddey01/switfs-auth/core"
)

var tokenSubCommand = &cli.Command{
	Name:  "token",
	Usage: "token command",
	Subcommands: []*cli.Command{
		genTokenCmd,
		getTokenCmd,
		listTokensCmd,
		removeTokenCmd,
		recoverTokenCmd,
	},
}

var genTokenCmd = &cli.Command{
	Name:      "gen",
	Usage:     "generate token",
	ArgsUsage: "[name]",
	UsageText: "./sophon-auth token gen --perm=<auth> [name]",
	Flags: []cli.Flag{
		&cli.StringFlag{
			Name:  "perm",
			Usage: "permission for API auth (read, write, sign, admin)",
		},
		&cli.StringFlag{
			Name:  "extra",
			Usage: "custom string in JWT payload",
			Value: "",
		},
	},
	Action: func(ctx *cli.Context) error {
		client, err := GetCli(ctx)
		if err != nil {
			return err
		}
		if ctx.NArg() < 1 {
			return fmt.Errorf("usage: genToken name")
		}
		name := ctx.Args().Get(0)

		if !ctx.IsSet("perm") {
			return fmt.Errorf("`perm` flag not set")
		}

		perm := ctx.String("perm")
		if !core.IsValid(perm) {
			return fmt.Errorf("`perm` flag invalid")
		}

		extra := ctx.String("extra")
		tk, err := client.GenerateToken(ctx.Context, name, perm, extra)
		if err != nil {
			return err
		}

		fmt.Printf("generate token success: %s\n", tk)
		return nil
	},
}

var getTokenCmd = &cli.Command{
	Name:  "get",
	Usage: "get token",
	Flags: []cli.Flag{
		&cli.StringFlag{
			Name: "name",
		},
		&cli.StringFlag{
			Name: "token",
		},
	},
	Action: func(ctx *cli.Context) error {
		name := ctx.String("name")
		token := ctx.String("token")
		client, err := GetCli(ctx)
		if err != nil {
			return err
		}
		tokens, err := client.GetToken(ctx.Context, name, token)
		if err != nil {
			return err
		}
		for _, token := range tokens {
			fmt.Println("name:       ", token.Name)
			fmt.Println("perm:       ", token.Perm)
			fmt.Println("create time:", token.CreateTime)
			fmt.Println("token:      ", token.Token)
			fmt.Println()
		}

		return nil
	},
}

var listTokensCmd = &cli.Command{
	Name:  "list",
	Usage: "list token info",
	Flags: []cli.Flag{
		&cli.UintFlag{
			Name:  "skip",
			Value: 0,
		},
		&cli.UintFlag{
			Name:  "limit",
			Value: 20,
			Usage: "max value:100 (default: 20)",
		},
	},
	Action: func(ctx *cli.Context) error {
		client, err := GetCli(ctx)
		if err != nil {
			return err
		}
		skip := int64(ctx.Uint("skip"))
		limit := int64(ctx.Uint("limit"))
		tks, err := client.Tokens(ctx.Context, skip, limit)
		if err != nil {
			return err
		}
		//	Token     string    `json:"token"`
		//	Name      string    `json:"name"`
		//	CreatTime time.Time `json:"createTime"`
		fmt.Println("num\tname\t\tperm\t\tcreateTime\t\ttoken")
		for k, v := range tks {
			name := v.Name
			if len(name) < 8 {
				name = name + "\t"
			}
			fmt.Printf("%d\t%s\t%s\t%s\t%s\n", k+1, name, v.Perm, v.CreateTime.Format("2006-01-02 15:04:05"), v.Token)
		}
		return nil
	},
}

var removeTokenCmd = &cli.Command{
	Name:      "rm",
	Usage:     "remove token",
	ArgsUsage: "[token]",
	Action: func(ctx *cli.Context) error {
		if ctx.NArg() < 1 {
			return fmt.Errorf("usage: rmToken [token]")
		}
		client, err := GetCli(ctx)
		if err != nil {
			return err
		}
		tk := ctx.Args().First()
		err = client.RemoveToken(ctx.Context, tk)
		if err != nil {
			return err
		}
		fmt.Printf("remove token success: %s\n", tk)
		return nil
	},
}

var recoverTokenCmd = &cli.Command{
	Name:      "recover",
	Usage:     "recover deleted token",
	ArgsUsage: "[token]",
	Action: func(ctx *cli.Context) error {
		if ctx.NArg() < 1 {
			return fmt.Errorf("usage: recover [token]")
		}
		client, err := GetCli(ctx)
		if err != nil {
			return err
		}
		tk := ctx.Args().First()
		err = client.RecoverToken(ctx.Context, tk)
		if err != nil {
			return err
		}
		fmt.Printf("recover token success: %s\n", tk)
		return nil
	},
}
